### PR TITLE
Added batch operation support

### LINKED ISSUE #52: Add batch operation support
## Details:
Hours Spent: 22

## Background
Member of epic #46. Very complex functions require a high amount of precision and may be very slow to compute. Hence the computation should be broken down into batches so that the user can see the view update with each batch that is processed.

## Overview
Add a new type `BatchOperation` that defines a plot region to compute which is bounded in x. Then add an array of these operations to the state. When the user clicks plot the expression should be broken into these batch commands. These commands are then sequentially executed. Inbetween each execution the view should be updated.

image

**Dev Note**
The sequential part should be done by having a computation action. This action should be triggered when new batch operations are added. It should take the first batch command and execute it asynchronously. When it is complete the view state should be updated and that batch command should be removed. Once the state is updated the action should be triggered again.

## Acceptance Criteria

- [x] Expressions are computed as batches
- [x] There is evidence that it performs as expected with an expression that is slow to compute (This can be done by delaying the expression computation).

## Closing Notes

Overview of changes can be found in PR #61

### PR DESCRIPTION
**Pull request by RelativeForce**
# Issues
closes #52 

# Summary of changes
- Disabled plot label drawing (Added issue #62 to address this)
- Refactored `Main` into a component `mainComponent` and broke its functionality into submodules
- Added `Queue` as an abstraction of `Data.List` which defines a basic queue interface
- Added `Id` type as an alias for `Int`
- Added `JobBatcher` which manages the `Job` queue. A `Job` is a `PlotCommand` with an `Id` and a batch `Id`. Each job has a unique `Id` and the batch `Id` is the `Id` the of the specific expression plot. `AddPlot` will break a robust plot into a number of plot commands that can be executed independently.
- Refactored the main component so that it will run the first job on the queue and draw it. Then it runs the next job in the queue recursively placing the next iteration on the end of the JS event queue.

# Tests
- Test.Plot.JobBatcher.CancelAll (Added)
- Test.Plot.JobBatcher.CancelWithBatchId (Added)
- Test.Plot.JobBatcher.ClearCancelled (Added)
- Test.Plot.JobBatcher.HasJobs (Added)
- Test.Plot.JobBatcher.InitialJobQueue (Added)
- Test.Plot.JobBatcher.IsCancelled (Added)
- Test.Plot.JobBatcher.SetRunning (Added)
- Test.Plot.JobBatcher.AddPlot (Added)

### CODE REVIEW

File: test/Plot/JobBatcher/JobBatcher.purs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Test.Expression.JobBatcher
```
Comment (michalkonecny) [inline test/Plot/JobBatcher/JobBatcher.purs]: The namespace of the module seems wrong.

**Comment by michalkonecny** [comment]
Did not manage to see it all today.  I will continue tomorrow.  So far it looks good to me.  
I think we should add temporary controls for number or segments and the size of batch to allow us to investigate the scalability of this approach.  I may try to add it myself over the weekend in a separate branch.

File: test/Plot/JobBatcher/JobBatcher.purs (hunk 1 above)
Comment (RelativeForce) [reply test/Plot/JobBatcher/JobBatcher.purs]: Resolved in latest commit

**Review by michalkonecny — APPROVED** [review]
All seems good.

**Pull request merged**